### PR TITLE
KAFKA-15150: Add ServiceLoaderScanner implementation

### PR DESCRIPTION
**Pull request by gharris1727**
Add the ServiceLoaderScanner, a companion to the ReflectionScanner which does not use reflection to discover plugins.

This will be utilized later in the Plugins class for the configurable-discovery-mode startup scanning, and in the connect-plugin-path script for manifest generation. In this PR, the PluginScannerTest is parameterized to ensure that the ServiceLoaderScanner has parity on the TestPlugins. This required adding manifests for the valid TestPlugins.

Also fixes a bug where the ServiceLoader handles plugin LinkageErrors differently than Reflections. To summarize:

Before:
* Reflections WARN logs LinkageError internally
* ServiceLoader propagates LinkageError and kills the worker

After:
* Reflections behaves the same as before
* JREs with ServiceLoader implementations that make progress before throwing LinkageError:
    * PluginScanner will typically log one error per unique error message, potentially up to 100 errors.
    * If there are >100 consecutive plugins failing with LinkageError, PluginScanner propagates the LinkageError and kills the worker.
* JREs with ServiceLoader implementations that do not make progress before throwing a LinkageError:
    * PluginScanner will typically log a single error, potentially up to 100 errors.
    * PluginScanner will propagate the LinkageError and kills the worker.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by gharris1727** [comment]
@C0urante Could you review this? Thanks!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/ReflectionScanner.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/ReflectionScanner.java]: Nit: can link to the KIP from here
```suggestion
 * implementation, consider using {@link ServiceLoaderScanner} and follow migration instructions for
 * <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-898%3A+Modernize+Connect+plugin+discovery">KIP-898</a>.
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java]: Nit:
```suggestion
    private final PluginScanner scanner;
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java]: I was wondering about replacing this part with a more rigorous assertion:
```java
Set<String> expectedClasses = new HashSet<>(TestPlugins.pluginClasses());
assertEquals(expectedClasses, classes);
```

But this fails with reflection-based scanning (and succeeds with service loader-based scanning) because the `VersionMethodThrowsConnector` class gets picked up by the plugin scanner, but isn't included in the result of `TestPlugins::pluginClasses`.

If we add a service loader manifest for the `VersionMethodThrowsConnector` class, then both cases fail. And then, if we change `includeByDefault` for the `BAD_PACKAGING_VERSION_METHOD_THROWS_CONNECTOR` to `true`, both test cases pass.

Should we add these changes? They give us slightly better coverage and more helpful assertion messages, and also align the total set of loaded classes when using each type of scanner (which IIUC matches the ideal scenario that we're aiming for once users have finished using the [KIP-898 migration script](https://cwiki.apache.org/confluence/display/KAFKA/KIP-898%3A+Modernize+Connect+plugin+discovery#KIP898:ModernizeConnectplugindiscovery-PluginPathManagementScript)).

**Review by C0urante — COMMENTED** [review]
Thanks Greg! A couple of nits and one question about testing methodology, then this should be good to go.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java
Comment (gharris1727) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java]: Thanks, I added both of your suggested fixes. The missing manifest and incorrect includeByDefault were oversights on my part, so the stronger test is a good idea.

Your suggestion also made me realize that I missed some ServiceLoader entries, meaning that ServiceLoaderScanner tests were incomplete. After I added the entries, two tests fail with an error (not an assertion) because the ServiceLoaderScanner doesn't catch the NoClassDefFound (LinkageError) that the ReflectionScanner does.

I added the failing MissingSuperclassConverter and NonExistentInterface test case in KAFKA-14649 #13182, to assert that they no longer crashed the worker, which I thought was reasonable at the time. It appears that the ServiceLoader takes the opinion that LinkageError (like other Errors) should not be caught due to the severity of the cause, which is making me second-think catching it in the ReflectionScanner.

I think there are some options here:
1. Delete the MissingSuperclassConverter manifest to disable the test case for the ServiceLoaderScanner
2. Catch the LinkageError in ServiceLoaderScanner making it shadow other plugins in the same PluginLocation.
3. Make ServiceLoaderScanner behave like ReflectionScanner by forking the ServiceLoader implementation and changing one catch clause.
4. Make ReflectionScanner behave like ServiceLoaderScanner and kill the worker for missing dependencies for plugin classes, restoring the pre-KAFKA-14649 behavior for LinkageError.

Of those options, I think i'm going to implement (4), as i'm loathe to fork ServiceLoader this early in the project, and the severity and breadth of causes for LinkageError means we probably shouldn't handle it.

What do you think? I've pushed the MissingSuperclassConverter manifest so the PluginScannerTest is currently failing.

Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java]: Of the approaches you've laid out, I'm -1 on 1 (the coverage provided by these cases is important, especially if it leads to discoveries like this one) and 2 (if a user runs into this it's going to be a massive headache to debug since they'll see plugin loading errors for more than just the broken plugin).

I agree that it's a bit early to go about implementing our own service loading logic, so although 3 is reasonable in theory, we can hold off on it for now.

I don't love 4, since it's still a step backwards and killing workers on startup is quite extreme. I think I could live with it if we believe there are no better alternatives.

I did some digging and it seems like this is due to a known bug in the OpenJDK `ServiceLoader`: https://bugs.openjdk.org/browse/JDK-8196182 (unclear if this also affects other JDK distributions).

Given that we're dealing with buggy behavior in a popular JDK distribution, it seems reasonable to try deal with this scenario gracefully. I prototyped a local workaround for this that's a little hacky but does cause our tests to pass. We can wrap calls to `Iterator::hasNext` in some error-handling logic:

```java
private boolean serviceLoaderHasNext(Iterator<?> serviceLoaderIterator) {
    try {
        return serviceLoaderIterator.hasNext();
    } catch (LinkageError e1) {
        log.error("Failed to scan for next service loaded plugin", e1);
        try {
            return serviceLoaderIterator.hasNext();
        } catch (LinkageError e2) {
            // It's difficult to know for sure if the iterator was able to advance past the first broken
            // plugin class, or if it will continue to fail on that broken class for any subsequent calls
            // to Iterator::hasNext or Iterator::next
            // As a best-effort measure, we compare exception messages with the assumption that
            // they will include the name of the plugin class; if the messages are the same, then we
            // fail fast and throw the original linkage error; if the messages differ, we try again
            // For reference, see https://bugs.openjdk.org/browse/JDK-8196182, which describes
            // the behavior we are trying to mitigate with this logic as buggy, but indicates that a fix
            // in the JDK standard library ServiceLoader implementation is unlikely to land
            if (Objects.equals(e1.getMessage(), e2.getMessage())) {
                throw e1;
            } else {
                log.error("Failed to scan for next service loaded plugin", e2);
                return true;
            }
        }
    }
}
```

Which would then change our for-loop in `PluginScanner::getServiceLoaderPluginDesc` to this:

```java
Iterator<T> iterator = serviceLoader.iterator();
while (serviceLoaderHasNext(iterator)) {
```

This essentially implements option 4, with some logic to try to intercept linkage errors and handle them gracefully before falling back on terminating the worker if it appears that our service loader is damaged beyond repair.

Thoughts?

Comment (gharris1727) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java]: I looked more into how the Reflections library handles this, and it actually just WARN logs these classes and never shows them to us, so we don't even get the opportunity to log the error ourselves:
```
[2023-07-14 11:54:26,418] WARN could not get type for name test.plugins.MissingSuperclassConverter from any class loader (org.reflections.Reflections:318)
org.reflections.ReflectionsException: could not get type for name test.plugins.MissingSuperclassConverter
	at org.reflections.ReflectionUtils.forName(ReflectionUtils.java:312)
	at org.reflections.ReflectionUtils.lambda$forNames$22(ReflectionUtils.java:330)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1621)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.reflections.ReflectionUtils.forNames(ReflectionUtils.java:332)
	at org.reflections.Reflections.getSubTypesOf(Reflections.java:404)
	at org.apache.kafka.connect.runtime.isolation.ReflectionScanner.getPluginDesc(ReflectionScanner.java:118)
	at org.apache.kafka.connect.runtime.isolation.ReflectionScanner.scanPlugins(ReflectionScanner.java:91)
	at org.apache.kafka.connect.runtime.isolation.PluginScanner.scanUrlsAndAddPlugins(PluginScanner.java:78)
	at org.apache.kafka.connect.runtime.isolation.PluginScanner.discoverPlugins(PluginScanner.java:66)
Caused by: java.lang.NoClassDefFoundError: test/plugins/NonExistentInterface
```

I think this takes option 4 off the table, as the reflection-loaded plugins have never failed due to LinkageErrors, before or after KAFKA-14649 change. The service-loaded plugins have always failed due to LinkageErrors.

I really like your implementation; it allows ServiceLoaders which make progress during exceptional hasNext calls to proceed to find other implementations and prevent shadowing. I mistakenly thought that it was iterator() throwing the exception instead of hasNext. Good catch! I'll tweak it a little to make it accommodate multiple plugins failing with LinkageErrors before triggering fallback behavior.

And I think I can accept re-throwing the error as a fallback behavior for ServiceLoaderScanner, if it only appears in situations where:
1. The JVM's implementation of ServiceLoader's Iterator's hasNext does not make progress on LinkageError
2. A plugin is installed which throws a LinkageError
3. plugin.discovery is in HYBRID_WARN, HYBRID_FAIL, or SERVICE_LOAD.

The first two conditions should be exceedingly rare, the overall situation has three different possibilities for remediation if it appears in production, and it allows us to still use the upstream ServiceLoader implementation. I'm sold.

Comment (gharris1727) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java]: Oh and just to motivate the tweaks I made to your implementation:
If the ServiceLoader implementation can skip over some LinkageErrors, I think it's possible to have two different plugins throw the same exception and accidentally trigger the equals condition.

For example, if two plugins implemented NonExistentInterface and appeared consecutively in the ServiceLoader manifest, I think the logic should show at least one of the errors, but not fail the worker or shadow other plugins. I think this would typically happen if a plugin packages both a Source and Sink that both have the same missing dependency, but the ServiceLoader implementation could continue to make progress.

There's a heuristic for 100 consecutive failures: if you package 100 consecutive faulty plugins (or have them all on the classpath) then it falls back to failing the worker. I don't think i've seen more than ~5-10 connectors, and ~20-30 Transforms packaged together. If the errors are non-consecutive then the counter resets as well, so the actual number of tolerated hasNext calls could be quite large, covering most use-cases.

The heuristic is just there to prevent infinite loops in the ServiceLoader-not-making-progress case. If the exception message happens to contain a memory address, or the exception alternates between two different messages, we should still fail the worker.

**Comment by gharris1727** [comment]
This appears to have some JDK8 failures that look consistent that i'll have to replicate locally. I've been testing with JDK11, so it's possible that the JDK8 ServiceLoader has some slightly different behavior, since it was overhauled in JDK9.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java]: This all looks great, thanks 👍

One small request for clarification: to me, this implies that we'd fail the worker without trying 100 times to invoke `hasNext`:

> If the exception message happens to contain a memory address, or the exception alternates between two different messages, we should still fail the worker.

But the current implementation still goes through the 100 tries before giving up. I agree with the implementation (better to err on the side of not unnecessarily killing the worker), but just wanted to double check--this is intentional, right?

Also, I notice that we've added `LinkageError` to the `catch` clause around invoking `Iterator::next`. Do we think it's worth adding similar logic to prevent infinite loops in the event that a call to `Iterator::next` that throws an exception also fails to advance the iterator? Or is this unreasonable paranoia?

**Comment by C0urante** [comment]
Tested the latest commit locally with OpenJDK 8 and 17; everything passes fine. Left one last comment but once that's addressed, this is good to go. Thanks for the diligent work on this, Greg!

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java
Comment (gharris1727) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java]: > But the current implementation still goes through the 100 tries before giving up. I agree with the implementation (better to err on the side of not unnecessarily killing the worker), but just wanted to double check--this is intentional, right?

Yes it is intentional. This edit should be more clear:

> The heuristic is just there to prevent infinite loops in the ServiceLoader-not-making-progress case.  If the exception message happens to contain a memory address, or the exception alternates between two different messages, we should still **eventually** fail the worker **and not be stuck in an infinite loop**.

> Also, I notice that we've added LinkageError to the catch clause around invoking Iterator::next. Do we think it's worth adding similar logic to prevent infinite loops in the event that a call to Iterator::next that throws an exception also fails to advance the iterator? Or is this unreasonable paranoia?

I suppose if we're being paranoid about hasNext, being paranoid about next() for JDK8 implementations would also make sense. I've made the error-handling function more generic and applied it to all of the ServiceLoader operations (including load and iterator, which i've never seen fail)

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java]: Nit: "resolve error within Converter plugin" doesn't seem very useful for users. Would it be possible to provide them with the plugin location instead?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java]: We added the `handleLinkageError` method because it was possible that calls to `Iterator::hasNext` and `Iterator::next` might skip over bad plugins, even if they threw a `LinkageError`. Do we have any reason to expect the same with `ServiceLoader::load` or `ServiceLoader::iterator`? If not, IMO we should leave these calls unwrapped.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java:139]: We provide a null `location` for the `PluginSource` corresponding to the system classpath, which won't look great in these log messages.

Since that field (and its accessor method) are currently only used for log messages, what do you think about altering `PluginSource::location` to return a string, and using "System classpath" in that case?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java]: Nit: I still think it's a bit misleading to specify the plugin class here. There may be a ton of broken plugins in this directory and if a user fixes all sink connectors and then sees this same error again telling them to fix all converters, they might rightfully be annoyed.

I think it's fine to just say "resolve errors for plugins in $PLUGIN_LOCATION" and leave it at that.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java]: Since we'll be silently enabling service loader scanning by setting the default to `HYBRID_WARN`, can we explicitly instruct users on how to disable service loader scanning (i.e., set `plugin.discovery` to `ONLY_SCAN`)?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/ReflectionScanner.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/ReflectionScanner.java]: Is this change intentional? The comma made sense IMO (though technically I suppose it should be a semicolon).

**Review by C0urante — COMMENTED** [review]
Thanks Greg. I really want to merge this but have some final thoughts. I promise this is it, though!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java]: In the case where an Implementation of ServiceLoader may eagerly-evaluate a first provider and cause a LinkageError to appear, I wanted to re-use the handleLinkageError logging.

I don't think that the retries for these operations would be effective, and that the later retries would probably be wasteful. However, I thought re-using the same function in both places was simpler than customizing the behavior of each call. Since this code is (probably never) going to be exercised, I tried to keep it as simple as possible.

Do you think that we should log errors from load and iterator? I can catch the error or just let it propagate to the caller.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java:139]: > Since that field (and its accessor method) are currently only used for log messages, what do you think about altering PluginSource::location to return a string, and using "System classpath" in that case?

I can't do that, because I need the Path object later in the migration script. I've used a non-null sentinel path instead, so that all of the logging call-sites are improved but the Path object is still available.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java]: This code-path is used even in ONLY_SCAN, so this remediation advice is actually misplaced. I'll remove this, and make sure that later when we implement the plugin.discovery mode that this error prints a separate recommendation to reconfigure the worker.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java]: I think that given the low likelihood of ever hitting this error when invoking `ServiceLoader::load` or `ServiceLoader::iterator` (neither is annotated with any kind of `throws` declaration, the iterator is specifically called out as being lazy-loading, and I'm not aware of any JDK bug reports that would lead to this kind of behavior), it's better to throw errors to the caller.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java:139]: Gotcha! The latest patch works, thanks.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java]: Great catch, thanks!

**Review by C0urante — APPROVED** [review]
LGTM (pending CI build). Thanks Greg, this was a tricky one!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java]: I've removed these usages, so the error propagates normally.

**Comment by gharris1727** [comment]
Flaky test failures appear unrelated.

**Pull request merged**
